rewriter/Prop.cc

```diff
@@ -255,12 +330,21 @@ vector<unique_ptr<ast::Expression>> processProp(core::MutableContext ctx, const
             loc, ast::MK::Hash1(loc, ast::MK::Symbol(nameLoc, core::Names::arg0()), ASTUtil::dupType(setType.get())),
             ASTUtil::dupType(setType.get())));
 
-        if (ret.enum_ == nullptr && forTStruct) {
-            // TODO(jez) Handle soft_freeze_logic when we implement non-T::Struct support
-            auto ivarSet = ast::MK::Assign(loc, ast::MK::Instance(nameLoc, ivarName),
-                                           ast::MK::Local(nameLoc, core::Names::arg0()));
-            nodes.emplace_back(ASTUtil::mkSet(ctx, loc, setName, nameLoc, std::move(ivarSet)));
+        if (ret.enum_ == nullptr && knownNonDocument(syntacticSuperClass)) {
+            if (wantTypedInitialize(syntacticSuperClass)) {
+                auto ivarSet = ast::MK::Assign(loc, ast::MK::Instance(nameLoc, ivarName),
+                                               ast::MK::Local(nameLoc, core::Names::arg0()));
+                nodes.emplace_back(ASTUtil::mkSet(ctx, loc, setName, nameLoc, std::move(ivarSet)));
+            } else {
+                // need to hide the instance variable access, because there wasn't a typed constructor to declare it
+                auto castSelf = ast::MK::Unsafe(loc, ast::MK::Self(loc));
+                auto ivarSet =
+                    ast::MK::Send2(loc, std::move(castSelf), core::Names::instanceVariableSet(),
+                                   ast::MK::Symbol(nameLoc, ivarName), ast::MK::Local(nameLoc, core::Names::arg0()));
+                nodes.emplace_back(ASTUtil::mkSet(ctx, loc, setName, nameLoc, std::move(ivarSet)));
+            }
         } else {
+            // Chalk::ODM::Document classes have special handling for soft freeze
```

Comment:
This is the second half of the change. For setters, we can only synthesize a body for `T::Struct` and `T::InexactStruct`, because other classes (like `Chalk::ODM::Document` and models) also do some `soft_freeze_logic` checking that I'll add in a future PR.